src/modules/spin.rs

```diff
@@ -0,0 +1,98 @@
+use regex::Regex;
+
+use crate::{
+    config::ModuleConfig,
+    configs::spin::SpinConfig,
+    context::Context,
+    formatter::{StringFormatter, VersionFormatter},
+    module::Module,
+};
+
+use crate::utils::get_command_string_output;
+const SPIN_CANARY: &str = "canary";
+const SPIN_CANARY_IDENTIFIER: &str = "pre";
+const SPIN_BINARY: &str = "spin";
+
+pub fn module<'a>(context: &'a Context) -> Option<Module<'a>> {
+    let mut module = context.new_module("spin");
+    let config = SpinConfig::try_load(module.config);
+    if !config.permanent {
+        let is_spin_project = context
+            .try_begin_scan()?
+            .set_files(&config.detect_files)
+            .set_extensions(&config.detect_extensions)
+            .set_folders(&config.detect_folders)
+            .is_match();
+
+        if !is_spin_project {
+            return None;
+        }
+    }
+
+    let parsed = StringFormatter::new(config.format).and_then(|formatter| {
+        formatter
+            .map_meta(|var, _| match var {
+                "symbol" => Some(config.symbol),
+                _ => None,
+            })
+            .map_style(|variable| match variable {
+                "style" => Some(Ok(config.style)),
+                _ => None,
+            })
+            .map(|variable| match variable {
+                "version" => {
+                    let spin_version_stdout =
+                        get_command_string_output(context.exec_cmd(SPIN_BINARY, &["--version"])?);
+                    let spin_version = parse_spin_version(spin_version_stdout.as_str())?;
+
+                    VersionFormatter::format_module_version(
+                        module.get_name(),
+                        &spin_version,
+                        config.version_format,
+                    )
+                    .map(Ok)
+                }
+                _ => None,
+            })
+            .parse(None, Some(context))
+    });
+
+    module.set_segments(match parsed {
+        Ok(segments) => segments,
+        Err(error) => {
+            log::warn!("Error in module `spin`:\n{}", error);
+            return None;
+        }
+    });
+
+    Some(module)
+}
+
+fn parse_spin_version(spin_version_stdout: &str) -> Option<String> {
+    let semver_regex = Regex::new(r"\d+\.\d+\.\d+").ok()?;
+    let input = spin_version_stdout;
+    let capture = semver_regex.find(input)?.as_str();
+    if input.contains(SPIN_CANARY_IDENTIFIER) {
+        Some(format!("{} ({})", capture, SPIN_CANARY))
+    } else {
+        Some(capture.to_string())
+    }
```

Comment:
Why do you replace `pre` in the version name with `canary`?

Reply:
Initially I thought it would be more self-explaining. However, you're right, `pre (2.8.0)` makes more sense than `canary (2.8.0)`

Reply:
I am not sure if rewriting the version here at all is a good idea.

Reply:
Okay, I changed the implementation to use versions provided by `spin --version` as they are